File: frontend/src/types/system-scheme.ts
import type { BuildingZone, DaySchedule } from '@/types/building'

// System scheme types — 项目级方案 + 子系统层级 (post-restructure)

export type SubsystemType = 'chiller_plant' | 'air_cooled' | 'shared_tower'
export type ConnectionType = 'direct' | 'parallel'
export type PipeSystem = 'two_pipe' | 'four_pipe'
export type StrategyStep = 'selection' | 'strategy' | 'diagram'
export type LoadDistributionMode = 'fixed_ratio' | 'by_priority' | 'other'
export type StrategyProfileMode = 'fixed' | 'by_month' | 'by_load' | 'by_dry_bulb' | 'by_wet_bulb' | 'constant_pressure'

// ---------------- Combo & TowerGroup ----------------

export interface SchemeCombo {
  id?: string
  combo_index: number
  primary_model_id: string | null
  primary_count: number
  primary_factor: number
  group_count: number
  chw_pump_model_id: string | null
  chw_pump_count: number
  chw_pump_backup: number
  chw_connection: ConnectionType
  chw_pump_factor: number
  cw_pump_model_id: string | null
  cw_pump_count: number
  cw_pump_backup: number
  cw_connection: ConnectionType
  cw_pump_factor: number
  extra?: Record<string, unknown> | null
}

export interface SchemeTowerGroup {
  id?: string
  group_index: number
  tower_model_id: string | null
  count: number
  factor: number
}

// ---------------- Design params ----------------

export interface ChillerPlantDesignParams {
  chw_supply_temp: number
  chw_delta_temp: number
  chw_pump_head: number
  header_pressure_drop: number
  cw_supply_temp: number
  cw_delta_temp: number
  cw_pump_head: number
}

export interface AirCooledDesignParams {
  cooling_supply_temp: number
  cooling_delta_temp: number
  heating_supply_temp: number
  heating_delta_temp: number
  pipe_system: PipeSystem
  pump_head?: number
  header_pressure_drop?: number
  chw_pump_head?: number
  chw_header_pressure_drop?: number
  hw_pump_head?: number
  hw_header_pressure_drop?: number
}

// ---------------- Subsystem ----------------

export interface Subsystem {
  id?: string
  scheme_id?: string
  subsystem_index: number
  subsystem_type: SubsystemType
  name: string
  design_params: Record<string, unknown>
  combos: SchemeCombo[]
  tower_groups: SchemeTowerGroup[]
}

// ---------------- Control strategy ----------------

export interface StrategyValueProfile {
  mode: StrategyProfileMode
  fixed_value: number
  month_values: number[]
  load_values: number[]
  dry_bulb_values: number[]
  wet_bulb_values: number[]
  constant_pressure: boolean
}

export interface LoadDistributionGroup {
  id: string
  name: string
}

export interface LoadDistributionGroupSetting {
  mode?: LoadDistributionMode
  ratios: Record<string, number>
  priorities: Record<string, number>
}

export interface LoadDistributionConfig {
  mode: LoadDistributionMode
  groups: LoadDistributionGroup[]
  zone_group_map: Record<string, string>
  subsystem_group_map: Record<string, string>
  group_settings: Record<string, LoadDistributionGroupSetting>
}

export interface StrategyStageBase {
  id: string
  combo_counts: Record<string, number>
  loading_down: number | null
  loading_up: number | null
  cooling_capacity_total: number
  cooling_capacity_min: number | null
  cooling_capacity_max: number | null
}

export interface ChillerStage extends StrategyStageBase {}

export interface AirCooledStage extends StrategyStageBase {
  heating_capacity_total: number
  heating_capacity_min: number | null
  heating_capacity_max: number | null
}

export interface PumpStrategy {
  min_freq: number
  max_freq: number
}

export interface TowerStrategy extends PumpStrategy {
  m: number
  k: number
}

export interface ChillerPlantStrategy {
  subsystem_id: string
  subsystem_type: 'chiller_plant'
  run_schedules: DaySchedule[]
  water_temp: {
    chw_supply: StrategyValueProfile
    chw_delta: StrategyValueProfile
    approach: StrategyValueProfile
    cw_delta: StrategyValueProfile
  }
  equipment: {
    chiller_stages: ChillerStage[]
    chw_pump: PumpStrategy
    cw_pump: PumpStrategy
    tower: TowerStrategy
  }
}

export interface AirCooledStrategy {
  subsystem_id: string
  subsystem_type: 'air_cooled'
  run_schedules: DaySchedule[]
  water_temp: {
    cooling_supply: StrategyValueProfile
    cooling_delta: StrategyValueProfile
    heating_supply: StrategyValueProfile
    heating_delta: StrategyValueProfile
  }
  equipment: {
    module_stages: AirCooledStage[]
    pump?: PumpStrategy
    chw_pump?: PumpStrategy
    hw_pump?: PumpStrategy
  }
}

export interface SharedTowerStrategy {
  subsystem_id: string
  subsystem_type: 'shared_tower'
  run_schedules: DaySchedule[]
  water_temp: Record<string, never>
  equipment: Record<string, never>
}

export type SubsystemControlStrategy = ChillerPlantStrategy | AirCooledStrategy | SharedTowerStrategy

export interface ControlStrategy {
  version: number
  load_distribution: LoadDistributionConfig
  system_strategies: Record<string, SubsystemControlStrategy>
}

export interface StrategyZoneSummary {
  key: string
  name: string
  area: number
  cooling_peak_est: number
  heating_peak_est: number
  source: BuildingZone
}

export interface StrategyGroupStat {
  group_id: string
  zone_count: number
  subsystem_count: number
  cooling_load_est: number
  heating_load_est: number
  cooling_capacity_total: number
  heating_capacity_total: number
}

// ---------------- Scheme (project-level) ----------------

export interface SystemScheme {
  id: string
  project_id: string
  building_id: string
  scheme_index: number
  name: string
  control_strategy: ControlStrategy
  diagram_json: Record<string, unknown>
  safety_margin: number
  subsystems: Subsystem[]
  created_at: string
  updated_at: string
}

export interface SystemSchemeCreate {
  name: string
  building_id: string
  scheme_index: number
  safety_margin?: number
  control_strategy?: ControlStrategy
  diagram_json?: Record<string, unknown>
  subsystems?: Subsystem[]
}

export type SystemSchemeUpdate = Partial<Omit<SystemSchemeCreate, 'building_id'>> & {
  building_id?: string
}

export interface SystemSchemeListItem {
  id: string
  name: string
  scheme_index: number
  project_id: string
  building_id: string
  building_name: string | null
  subsystem_count: number
  cooling_capacity_total: number
  heating_capacity_total: number
  cooling_load_peak: number | null
  heating_load_peak: number | null
  /** Annual cooling output (kWh). Null until energy simulation is run. */
  annual_cooling_total: number | null
  /** Annual heating output (kWh). */
  annual_heating_total: number | null
  /** Annual system energy consumption (kWh). */
  annual_energy_total: number | null
  /** System annual COP / efficiency. */
  system_cop: number | null
  /** Annual operating cost (CNY). */
  annual_cost: number | null
  has_error: boolean
  updated_at: string
}

// ---------------- Validation ----------------

export interface ValidationIssue {
  code: string
  severity: 'error' | 'warning'
  message: string
  scheme_id?: string | null
  subsystem_id?: string | null
  combo_id?: string | null
  field?: string | null
}

export interface ValidationReport {
  valid: boolean
  issues: ValidationIssue[]
}

export interface CapacitySummary {
  cooling_load_peak: number | null
  heating_load_peak: number | null
  cooling_capacity_total: number
  heating_capacity_total: number
}

// ---------------- Derived display data ----------------

export interface PumpView {
  brand: string | null
  name: string
  flow: number
  head: number
  power: number
  efficiency: number
  active_count: number
}

export interface SchemeDerivedCombo {
  id: string
  combo_index: number
  primary: { id: string; name: string; brand: string | null; model_no: string | null; series: string | null } | null
  primary_count: number
  group_count: number
  cooling_capacity: number
  heating_capacity: number
  power: number
  heating_power: number
  cop: number | null
  cop_heat: number | null
  chw_flow: number
  cw_flow: number
  evap_dp: number
  cond_dp: number
  chw_pump: PumpView | null
  cw_pump: PumpView | null
}

export interface SubsystemDerived {
  id: string
  subsystem_index: number
  subsystem_type: SubsystemType
  name: string
  cooling_capacity_total: number
  heating_capacity_total: number
  combos: SchemeDerivedCombo[]
  tower_groups: Array<{
    id: string
    group_index: number
    model: { id: string; name: string; brand: string | null; series: string | null } | null
    count: number
    flow: number
    power: number
    head: number
    inlet_temp: number
    outlet_temp: number
    wet_bulb: number
  }>
  tower_flow_total: number
}

export interface SchemeDerived {
  scheme_id: string
  cooling_capacity_total: number
  heating_capacity_total: number
  subsystems: SubsystemDerived[]
}

// ---------------- Equipment search ----------------

export interface EquipmentSearchParams {
  equipment_type: 'chiller' | 'air_cooled_module' | 'pump' | 'cooling_tower'
  name?: string
  capacity_min?: number
  capacity_max?: number
  flow_min?: number
  flow_max?: number
  head_min?: number
  head_max?: number
  efficiency_min?: number
  efficiency_max?: number
}

export interface EquipmentBrief {
  id: string
  name: string
  equipment_type: string
  brand: string | null
  model_no: string | null
  capacity: number | null
  cop: number | null
  parameters: Record<string, unknown> | null
  is_public: boolean
}

export const SUBSYSTEM_TYPE_LABELS: Record<SubsystemType, string> = {
  chiller_plant: 'scheme.types.chiller_plant',
  air_cooled: 'scheme.types.air_cooled',
  shared_tower: 'scheme.types.shared_tower',
}

File: backend/app/schemas/system_scheme.py
"""Pydantic schemas for system schemes (项目级方案 + 多子系统层级)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, ConfigDict


SubsystemType = Literal["chiller_plant", "air_cooled", "shared_tower"]
ConnectionType = Literal["direct", "parallel"]
PipeSystem = Literal["two_pipe", "four_pipe"]


# ---------- Combo ----------

class SchemeComboBase(BaseModel):
    id: uuid.UUID | None = None  # echoed back so dry-run validation can map issues to existing rows
    combo_index: int = Field(1, ge=1, le=10)
    primary_model_id: uuid.UUID | None = None
    primary_count: int = Field(1, ge=1, le=20)
    primary_factor: float = Field(0.92, ge=0.10, le=1.00)
    group_count: int = Field(1, ge=1, le=10)
    chw_pump_model_id: uuid.UUID | None = None
    chw_pump_count: int = Field(1, ge=1, le=20)
    chw_pump_backup: int = Field(0, ge=0, le=1)
    chw_connection: ConnectionType = "direct"
    chw_pump_factor: float = Field(0.77, ge=0.10, le=1.00)
    cw_pump_model_id: uuid.UUID | None = None
    cw_pump_count: int = Field(1, ge=1, le=20)
    cw_pump_backup: int = Field(0, ge=0, le=1)
    cw_connection: ConnectionType = "direct"
    cw_pump_factor: float = Field(0.77, ge=0.10, le=1.00)
    extra: dict[str, Any] | None = None


class SchemeComboResponse(SchemeComboBase):
    id: uuid.UUID
    model_config = ConfigDict(from_attributes=True)


# ---------- Tower group ----------

class SchemeTowerGroupBase(BaseModel):
    id: uuid.UUID | None = None  # echoed back so dry-run validation can map issues to existing rows
    group_index: int = Field(1, ge=1, le=10)
    tower_model_id: uuid.UUID | None = None
    count: int = Field(1, ge=1, le=20)
    factor: float = Field(0.85, ge=0.10, le=1.00)


class SchemeTowerGroupResponse(SchemeTowerGroupBase):
    id: uuid.UUID
    model_config = ConfigDict(from_attributes=True)


# ---------- Design parameters (per subsystem type, JSON) ----------

class ChillerPlantDesignParams(BaseModel):
    chw_supply_temp: float = Field(7.0, ge=1, le=25)
    chw_delta_temp: float = Field(5.0, ge=1, le=15)
    chw_pump_head: float = Field(35.0, ge=1, le=100)
    header_pressure_drop: float = Field(21.0, ge=1, le=100)
    cw_supply_temp: float = Field(30.0, ge=1, le=50)
    cw_delta_temp: float = Field(5.0, ge=1, le=15)
    cw_pump_head: float = Field(30.0, ge=1, le=100)


class AirCooledDesignParams(BaseModel):
    cooling_supply_temp: float = Field(7.0, ge=1, le=25)
    cooling_delta_temp: float = Field(5.0, ge=1, le=15)
    heating_supply_temp: float = Field(45.0, ge=30, le=100)
    heating_delta_temp: float = Field(5.0, ge=1, le=15)
    pipe_system: PipeSystem = "two_pipe"
    pump_head: float | None = Field(35.0, ge=1, le=100)
    header_pressure_drop: float | None = Field(21.0, ge=1, le=100)
    chw_pump_head: float | None = Field(35.0, ge=1, le=100)
    chw_header_pressure_drop: float | None = Field(21.0, ge=1, le=100)
    hw_pump_head: float | None = Field(35.0, ge=1, le=100)
    hw_header_pressure_drop: float | None = Field(21.0, ge=1, le=100)


# ---------- Subsystem ----------

class SubsystemBase(BaseModel):
    id: uuid.UUID | None = None  # echoed back so dry-run validation can map issues to existing rows
    subsystem_index: int = Field(1, ge=1, le=20)
    subsystem_type: SubsystemType
    name: str = Field(..., max_length=200)
    design_params: dict[str, Any] = Field(default_factory=dict)
    combos: list[SchemeComboBase] = Field(default_factory=list)
    tower_groups: list[SchemeTowerGroupBase] = Field(default_factory=list)


class SubsystemCreate(SubsystemBase):
    pass


class SubsystemUpdate(BaseModel):
    subsystem_index: int | None = Field(None, ge=1, le=20)
    name: str | None = Field(None, max_length=200)
    design_params: dict[str, Any] | None = None
    combos: list[SchemeComboBase] | None = None
    tower_groups: list[SchemeTowerGroupBase] | None = None


class SubsystemResponse(BaseModel):
    id: uuid.UUID
    scheme_id: uuid.UUID
    subsystem_index: int
    subsystem_type: str
    name: str
    design_params: dict[str, Any]
    combos: list[SchemeComboResponse] = Field(default_factory=list)
    tower_groups: list[SchemeTowerGroupResponse] = Field(default_factory=list)
    model_config = ConfigDict(from_attributes=True)


# ---------- SystemScheme ----------

class SystemSchemeCreate(BaseModel):
    name: str = Field(..., max_length=200)
    building_id: uuid.UUID
    scheme_index: int = Field(1, ge=1, le=5)
    safety_margin: float = Field(1.0, ge=0, le=1.2)
    control_strategy: dict[str, Any] = Field(default_factory=dict)
    diagram_json: dict[str, Any] = Field(default_factory=dict)
    subsystems: list[SubsystemCreate] = Field(default_factory=list)


class SystemSchemeUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    building_id: uuid.UUID | None = None
    scheme_index: int | None = Field(None, ge=1, le=5)
    safety_margin: float | None = Field(None, ge=0, le=1.2)
    control_strategy: dict[str, Any] | None = None
    diagram_json: dict[str, Any] | None = None
    subsystems: list[SubsystemCreate] | None = None  # full replace


class SystemSchemeResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    building_id: uuid.UUID
    scheme_index: int
    name: str
    control_strategy: dict[str, Any]
    diagram_json: dict[str, Any]
    safety_margin: float
    subsystems: list[SubsystemResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class SystemSchemeListItem(BaseModel):
    """Light-weight item for list cards on the project-level main page."""
    id: uuid.UUID
    name: str
    scheme_index: int
    project_id: uuid.UUID
    building_id: uuid.UUID
    building_name: str | None = None
    subsystem_count: int = 0
    safety_margin: float = 1.0
    cooling_capacity_total: float = 0.0
    heating_capacity_total: float = 0.0
    cooling_load_peak: float | None = None
    heating_load_peak: float | None = None
    # Annual energy simulation results (populated when energy simulation is run;
    # currently None until energy simulation module is implemented).
    annual_cooling_total: float | None = None  # kWh
    annual_heating_total: float | None = None  # kWh
    annual_energy_total: float | None = None  # kWh
    system_cop: float | None = None  # dimensionless
    annual_cost: float | None = None  # CNY
    has_error: bool = False
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ---------- Validation report ----------

class ValidationIssue(BaseModel):
    code: str
    severity: Literal["error", "warning"]
    message: str
    scheme_id: uuid.UUID | None = None
    subsystem_id: uuid.UUID | None = None
    combo_id: uuid.UUID | None = None
    field: str | None = None


class ValidationReport(BaseModel):
    valid: bool
    issues: list[ValidationIssue] = Field(default_factory=list)


# ---------- Capacity summary (per scheme) ----------

class CapacitySummary(BaseModel):
    cooling_load_peak: float | None = None
    heating_load_peak: float | None = None
    cooling_capacity_total: float = 0.0
    heating_capacity_total: float = 0.0


# ---------- Equipment search ----------

class EquipmentSearchQuery(BaseModel):
    equipment_type: str
    name: str | None = None
    capacity_min: float | None = None
    capacity_max: float | None = None
    flow_min: float | None = None
    flow_max: float | None = None
    head_min: float | None = None
    head_max: float | None = None
    efficiency_min: float | None = None
